File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(0, new ArrayList<>(), nums, result);
        return result;
    }

    private void dfs(int start, List<Integer> visited,
                     int[] nums, List<List<Integer>> result) {
        result.add(new ArrayList<>(visited));

        for (int idx = start; idx < nums.length; idx++) {
            visited.add(nums[idx]);
            dfs(idx + 1, visited, nums, result);
            visited.remove(visited.size() - 1);
        }
    }
}

File: leetcode/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function subsets(nums) {
    const result = [];

    function dfs(start, visited) {
        result.push([...visited]);

        for (let idx = start; idx < nums.length; idx++) {
            visited.push(nums[idx]);
            dfs(idx + 1, visited);
            visited.pop();
        }
    }

    dfs(0, []);

    return result;
}